docs/rfcs/033-timeline-archive.md

```diff
@@ -0,0 +1,291 @@
+be controlled by a new per-timeline `configure` endpoint.  This is intentionally generic naming, which
+may be used in future to control other per-timeline attributes (e.g. in future we may make PITR interval
+a per-timeline configuration).
+
+`PUT /v1/tenants/{tenant_id}/timelines/{timeline_id}/configure`
+```
+{
+  'state': 'active|archive'
+}
+```
+
+When archiving a timeline, this API will complete as soon as the timeline's state has been set in index_part, and that index has been uploaded.
+
+When activating a timeline, this API will complete as soon as the timeline's state has been set in index_part,
+**and** the `Timeline` object has been instantiated and activated.  This will require reading the timeline's
+index, but not any data: it should be about as fast as a couple of small S3 requests.
+
+The API will be available with identical path via the storage controller: calling this on a sharded tenant
+will simply map the API call to all the shards.
+
+### Tenant attach changes
+
+Currently, during Tenant::spawn we list all the timelines in the S3 bucket, and then for each timeline
+we load their index_part.json.  To avoid this operation scaling linearly with the number of archived
+timelines, we must have a single object that tells us which timelines do not need to be loaded.
+
+This is **not** literally the same as the set of timelines who have state=archived.  Rather, it is
+the set of timelines which have been offloaded in the background after their state was set to archived.
+
+We may simply skip loading these timelines: there will be no special state of `Timeline`, they just won't
+exist from the perspective of an active `Tenant` apart from in deletion: timeline deletion will need
+to check for offloaded timelines as well as active timelines, to avoid wrongly returning 404 on trying
+to delete an offloaded timeline.
+
+### Warm-up API
+
+`PUT /v1/tenants/{tenant_id}/timelines/{timeline_id}/download?wait_ms=1234`
+
+This API will be similar to the existing `download_remote_layers` API, but smarter:
+- It will not download _all_ remote layers, just the visible set (i.e. layers needed for a read)
+- It will download layers until reaching `wait_ms`, then return a struct describing progress
+  of downloads, so that the caller can poll.
+
+The caller does not have to wait for the warm up API, or call it at all.  But it is strongly advised
+to call it, because otherwise populating local contents for a timeline can take a long time when waiting
+for SQL queries to coincidentally hit all the layers, and during that time query latency remains quite
+volatile.
+
+### Background offload
+
+Recall that when we archive a timeline via the HTTP API, this only sets a state: it doesn't do
+any actual work.
+
+This work is done in the background compaction loop.  It makes sense to tag this work on to the compaction
+loop, because it is spiritually aligned: offloading data for archived branches improves storage efficiency.
```

Comment:
> adding more stuff to the compaction loop is an anti-pattern. We are working in an async runtime and tasks should be cheap

I'm thinking about this in terms of which things we want to run concurrently: I would put the "background storage optimization" stuff under the same semaphore we use to limit concurrent compaction, and with 10,000s of tenants on a pageserver the case of running compaction + storage optimization for the same tenant concurrently is kind of niche, so I lean toward minimizing LOC by just tucking it into the existing loop.

> perhaps we should rework that to allow for separate tasks (shard tenant-shards and spawn a fixed number of compaction tasks, where each task handles more than one tenant).

Right: the "spawn 10,000 tasks and have them fight for a semaphore" model is one I want to move away from, not because of machine efficiency (a semaphore is just a queue), but because the simple tasks+semaphore doesn't let us express relative priority of tenants.  I would like to eventually replaced the compaction task-per-tenant with something similar to (or derived from) the `scheduler.rs` used for heatmaps/secondary downloads, and order the tasks by how urgently a tenant needs to do compaction.

Reply:
> I would like to eventually replaced the compaction task-per-tenant with something similar to (or derived from) the scheduler.rs used for heatmaps/secondary downloads, and order the tasks by how urgently a tenant needs to do compaction.

+1 This is similar to what I had in mind, but didn't articulate very well.

> and with 10,000s of tenants on a pageserver the case of running compaction + storage optimization for the same tenant concurrently is kind of niche

Exactly. To my mind, doing storage optimisations _and_ compaction on archived timeline doesn't make sense. So the compaction loop becomes a bimodal thing. Maybe that's fine. It could be split up quite cleanly. Check if timeline is archived and decide whether to do compaction or storage optimisations based on that.

Unrelated note: can we call "storage optimisations" something else (maybe "archive optimisation"). It's a bit too generic for my liking.

Reply:
> Unrelated note: can we call "storage optimisations" something else (maybe "archive optimisation"). It's a bit too generic for my liking.

done